### ISSUE TITLE
--equal does not work on files with numerical uid or gid

### ISSUE THREAD

**Issue by hungrywolf27** [sahib/rmlint#234]
On Ubuntu Linux. `rmlint --equal` reports files as different if one of them has a numerical uid or gid that is not in `/etc/passwd`.

Let's say I am user `alice`, and user `bob` also exists on the system:

```
$ cp testfile file1
$ cp testfile file2
$ rmlint --equal file1 file2 ; echo $?
0
$ sudo chown bob:bob file1 
$ rmlint --equal file1 file2 ; echo $?
0
$ sudo chown bob:789 file1 
$ rmlint --equal file1 file2 ; echo $?
1
$ diff file1 file2 ; echo $?
0
$ cmp file1 file2 ; echo $?
0
$ sudo chown 789:bob file1 
$ rmlint --equal file1 file2 ; echo $?
1
$ cmp file1 file2 ; echo $?
0
$ diff file1 file2 ; echo $?
0
$ sudo chown bob:bob file1 
$ rmlint --equal file1 file2 ; echo $?
0
```

**Comment by hungrywolf27** [comment]
Aha, `--equal` works as expected if I add `--types=df`:
```
$ sudo chown 789:bob file1 
$ rmlint --equal file1 file2 ; echo $?
1
$ rmlint --equal --types=df file1 file2 ; echo $?
0
```

The problem is that when I run `rmlint --types=df [dir1 dir2 dir3 ...]`, and then run `rmlint.sh -p`, I get errors of "files no longer identical".

**Comment by sahib** [comment]
Apparently I found a quick and good way to get quality control: do a release :PYeah, ``rmlint`` also finds Files with broken uid/his permissions. ``--equal`` should limit itself to duplicate files and directories. I will fix that later.Am 09.06.2017 12:33 nachm. schrieb hungrywolf27 <notifications@github.com>:Aha, --equal works as expected if I add --types=df:
$ sudo chown 789:bob file1 
$ rmlint --equal file1 file2 ; echo $?
1
$ rmlint --equal --types=df file1 file2 ; echo $?
0

The problem is that when I run rmlint --types=df [dir1 dir2 dir3 ...], and then run rmlint.sh -p, I get errors of "files no longer identical".

**Comment by sahib** [comment]
Should be fixed now in the head of the ``develop`` branch. 
``--equal`` now limits itself to duplicate files and duplicate directories (``-T dd,df``).

Another change in the same direction: ``rmlint`` will also recognize two empty file as equal now (and see them as duplicates if we're only searching for duplicate files).

**Comment by hungrywolf27** [comment]
Wait a minute, I think it's a bad idea for `rmlint -T df` to remove empty files by default.

There's already a separate option for finding and removing empty files (`-T ef`), and if I'm not using it, I'd expect my empty files to be left alone.
http://rmlint.readthedocs.io/en/latest/rmlint.1.html#general-options

**Comment by sahib** [comment]
> Wait a minute, I think it's a bad idea for rmlint -T df to remove empty files by default.

Would you elaborate a bit on this?

> There's already a separate option for finding and removing empty files (-T ef), and if I'm not using it, I'd expect my empty files to be left alone.
> http://rmlint.readthedocs.io/en/latest/rmlint.1.html#general-options

Yes, I'm aware that it would be a change in the documented behavior. At least for ``--equal`` it makes sense to treat empty files as equal and from a pure logic point of view it makes sense to also recognize them as duplicates in normal mode (they have the same content: none). The reason why we treat empty files special is mostly for historic reasons and because *other tools did it* (e.g. ``fslint``). But overall that's not a good reason. 

I would be interested in hearing your use case.
(Note: you can still exclude empty files, it's just not the default in ``develop``: ``$ rmlint -T df --size 1``)

**Comment by hungrywolf27** [comment]
OK, at first I thought this change left no way to exclude empty files (since `-T df,-ef` no longer works). Thanks for pointing out `--size 1`. Here are the main reasons I think that `-T df` shouldn't remove empty files by default:
- It's a change in documented behavior.
- My experience with other dupe finders (fslint and rdfind) is that empty files are ignored by default, or treated as a separate category.
- Treating empty files as a separate category makes sense because the desired behavior is usually to delete _all_ empty files, but to keep at least one copy of non-empty files. (What is the point of keeping one copy of an empty file?)
- One could also take the perspective that having _no content at all_ is not the same as _having the same content_. Because empty files have no content, there is nothing to compare, and so they are not duplicates.
- Sometimes people create empty files with a filename as a "note to self" when browsing a folder, in which case the filename itself is effectively the content of the file. (Also, of course, a typical Linux home folder contains many zero-byte files in hidden directories used as lockfiles, etc., but these hidden directories would be excluded by default.)

**Comment by sahib** [comment]
> It's a change in documented behavior.

Yes, it is. But documentation is not set in stone and will be changed. Since this change is somewhat "breaking" it will surely be mentioned in the Changelog and highlighted in the documentation.

> My experience with other dupe finders (fslint and rdfind) is that empty files are ignored by default, or treated as a separate category.

As I said, that was the reason it was implemented that way. But *others do it* is not necessarily a good reason.

> Treating empty files as a separate category makes sense because the desired behavior is usually to delete all empty files, but to keep at least one copy of non-empty files. (What is the point of keeping one copy of an empty file?)

That's still the current behavior since by default ``rmlint`` will search for empty files by default.
The case only applies when using ``-T df`` (i.e. empty file finding disabled).

At the point of keeping empty files: As you said below: Some people define some semantics in the pure presence of empty files which ``rmlint`` of course does not know about. If the user wishes to only search for duplicates then empty files should be treated the same as duplicates. Again, the default behavior is more opinionated.

> One could also take the perspective that having no content at all is not the same as having the same content. Because empty files have no content, there is nothing to compare, and so they are not duplicates.

There are of course many possible definitions. Still, when designing a tool that checks for equality of files, I'd prefer a practical and useful definition that does not have any special cases that users need to watch for.

In the context of ``rmlint`` a comparison with hashing utilities might make sense:

```
$ touch empty_a empty_b
$ md5sum empty_a empty_b
d41d8cd98f00b204e9800998ecf8427e empty_a
d41d8cd98f00b204e9800998ecf8427e empty_b
```

Another point here: For ``--equal`` it IMHO definitely makes sense to make it report a ``0`` when comparing ``empty_a`` and ``empty_b`` (would you expect differently?). If now other parts of ``rmlint`` would decide
differently (i.e. not outputting empty "duplicates"), then this wouldn't strike me as very consistent.


> Sometimes people create empty files with a filename as a "note to self" when browsing a folder, in which case the filename itself is effectively the content of the file. (Also, of course, a typical Linux home folder contains many zero-byte files in hidden directories used as lockfiles, etc., but these hidden directories would be excluded by default.)

Valid points. As I said above ``rmlint`` cannot know about the semantics that some user defined. It always has to support both sides, which it does in this case. It's just a matter of deciding the most useful default for a majority of users. There will always be discussion who's part of the majority.... :)

I did not reach a conclusion yet, so I'll give that still a night or two to think about. Indeed that's the first time I'm thinking in deep about this behavior.

**Comment by SeeSpotRun** [comment]
I'm marginally on @hungrywolf27's side on this one.  Empty files are pure metadata and so deserve some sort of special handling.  Users wanting to remove "duplicate" empty files can always use `--size 0`.

But it makes sense for `rmlint --equal` to default to `--size 0`.

**Comment by sahib** [comment]
> I'm marginally on @hungrywolf27's side on this one. Empty files are pure metadata and so deserve some sort of special handling. Users wanting to remove "duplicate" empty files can always use --size 0.
> 
> But it makes sense for rmlint --equal to default to --size 0.

I could live with that, despite still favoring the approach I described up there.
But apparently I'm in the minority. :smile:

At least we won't confuse any users from other tools and do not need to adjust documentation (which, out of experience, nobody will read that, just because we released a new version).

To summarize and clearly state what will be done:

* Set ``--size`` default to ``1-inf``, thus not finding empty files as duplicates (was ``0-inf`` before).
* Set ``--size`` to ``0-inf`` when using ``--equal``.
* Add a note somewhere in the docs that ``--size 0`` will find empty file duplicates.
* Add a test case to check if that worked out.
